blog/_posts/2025-04-04-Traffic-Routing-Demystified-Best-practices-for-AKS-Load-Balancers.md

```diff
@@ -0,0 +1,141 @@
+
+In this blog, we delve into the intricacies of Kubernetes’ `externalTrafficPolicy=Local` setting and explore strategies to gracefully handle pod shutdowns, rolling updates, and pod distribution. By following these best practices, you can enhance the resilience and reliability of your services while optimizing resource utilization across your AKS clusters.
+
+## `externalTrafficPolicy`: Cluster vs. Local
+
+**What does `externalTrafficPolicy` do?** In short, it influences how a cloud load balancer directs incoming traffic to Kubernetes nodes and how that traffic may hop between nodes to reach pods:
+
+- **Cluster Mode (default) (`externalTrafficPolicy=Cluster`)**:
+    All nodes in the cluster are included behind the Azure Load Balancer (ALB), regardless of whether they host any pods for the service. Incoming external traffic is distributed evenly across all nodes in the cluster. If a node receives traffic but doesn't have a local pod for the Service, Kubernetes (`kube-proxy`) will forward the request across the cluster network to a node that does. While this mode simplifies traffic distribution and reduces the risk of uneven load, it does not preserve the client’s original source IP.
+
+- **Local Mode (`externalTrafficPolicy=Local`)**:
+    In this mode, external traffic is directed only to nodes that host healthy pods for the specified service. The traffic is routed exclusively to the pods residing on the same node where it is received. This approach ensures that the client’s original source IP is preserved, which is critical for use cases involving security, logging, and analytics. However, it requires careful consideration of pod distribution across nodes to avoid uneven traffic loads.
+    If one node hosts multiple pods while another node hosts only one, the node with one pod will receive roughly the same share of incoming connections as the node with many pods.
+
+By understanding the operational differences between these two modes, you can make an informed decision based on your application’s requirements for traffic routing, source IP preservation, and load balancing.
+
+### Benefits of Local Mode
+
+- **Localized Impact of Node Downtime**: Traffic is affected only if the downed node is running a service pod, impacting that pod’s share of the traffic.
+- **Preservation of Client Source IP**: The client’s original IP is maintained, which is crucial for security, logging, and analytics.
+- **Targeted Health Checking**: Local mode uses a dedicated `healthCheckNodePort`, ensuring that external load balancers send traffic only to nodes with healthy pods.
```

Comment:
This is true, but I don't know that it's a particular benefit.  The end result (traffic only goes to healthy pods) is ensured by either externalTrafficPolicy.

Reply:
agreed, could be cleaned up a bit to say the benefit is more efficient than where cluster might add extra hops to your traffic.